File: app/src/main/java/com/daohoangson/ydls/viewmodel/OpenGraphLiveData.java
package com.daohoangson.ydls.viewmodel;

import android.app.Application;
import android.os.Handler;
import android.os.Looper;
import android.text.TextUtils;
import android.view.View;
import android.widget.ImageView;
import android.widget.TextView;

import com.android.volley.NetworkResponse;
import com.android.volley.Request;
import com.android.volley.RequestQueue;
import com.android.volley.Response;
import com.android.volley.VolleyError;
import com.android.volley.toolbox.HttpHeaderParser;
import com.android.volley.toolbox.Volley;
import com.daohoangson.ydls.R;
import com.squareup.picasso.Picasso;

import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;
import org.jsoup.select.Elements;

import java.io.UnsupportedEncodingException;
import java.net.HttpURLConnection;
import java.util.HashMap;
import java.util.Map;

import androidx.core.content.ContextCompat;
import androidx.databinding.BindingAdapter;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.Observer;

public class OpenGraphLiveData extends LiveData<OpenGraphLiveData.Data> implements Observer<String> {

    private static final String TAG = "OpenGraph";

    private final Handler mHandler;
    private final RequestQueue mQueue;
    private final LiveData<String> mSourceMediaUrl;

    OpenGraphLiveData(Application application, LiveData<String> mediaUrl) {
        mHandler = new Handler(Looper.getMainLooper());

        mQueue = Volley.newRequestQueue(application.getApplicationContext());

        // https://github.com/google/volley/issues/51
        HttpURLConnection.setFollowRedirects(true);

        mSourceMediaUrl = mediaUrl;

        setValue(new Data(null));
    }

    @Override
    protected void onActive() {
        super.onActive();

        mSourceMediaUrl.observeForever(this);
    }

    @Override
    public void onChanged(final String url) {
        mHandler.postDelayed(new Runnable() {
            @Override
            public void run() {
                parseUrl(url);
            }
        }, 1000);
    }

    @Override
    protected void onInactive() {
        super.onInactive();

        cancelQueue();
        mSourceMediaUrl.removeObserver(this);
    }

    private void cancelQueue() {
        mQueue.cancelAll(TAG);
    }

    private void parseUrl(String url) {
        if (Looper.myLooper() != Looper.getMainLooper()) {
            throw new IllegalThreadStateException();
        }

        if (!url.equals(mSourceMediaUrl.getValue())) {
            // url is outdated (because of the delayed processing)
            return;
        }

        Data currentData = getValue();
        if (currentData != null) {
            if (url.equals(currentData.mediaUrl)) {
                // url is already being processed
                return;
            }

            if (currentData.isLoading) {
                cancelQueue();
            }
        }

        Data loadingData = new Data(url);
        loadingData.isLoading = true;
        postValue(loadingData);

        OgRequest req = new OgRequest(url);
        mQueue.add(req.setTag(TAG));
    }

    public static class Data {
        boolean hasData = false;
        boolean isLoading = false;
        final String mediaUrl;
        String title = "";
        String imageUrl = "";

        Data(String mediaUrl) {
            this.mediaUrl = mediaUrl;
        }
    }

    class OgRequest extends Request<Data> {
        private final String url;

        OgRequest(final String url) {
            super(
                    Method.GET,
                    url,
                    new Response.ErrorListener() {
                        @Override
                        public void onErrorResponse(VolleyError error) {
                            postValue(new Data(url));
                        }
                    }
            );

            this.url = url;
        }

        @Override
        protected void deliverResponse(Data d) {
            postValue(d);
        }

        @Override
        public Map<String, String> getHeaders() {
            Map<String, String> headers = new HashMap<>();

            headers.put("User-Agent", "libcurl");

            return headers;
        }

        @Override
        @SuppressWarnings("DefaultCharset")
        protected Response<Data> parseNetworkResponse(NetworkResponse response) {
            String html;
            try {
                html = new String(response.data, HttpHeaderParser.parseCharset(response.headers));
            } catch (UnsupportedEncodingException e) {
                // Since minSdkVersion = 8, we can't call
                // new String(response.data, Charset.defaultCharset())
                // So suppress the warning instead.
                html = new String(response.data);
            }

            Data d = new Data(url);
            d.hasData = true;

            Document doc = Jsoup.parse(html);
            Elements titleTags = doc.getElementsByTag("title");
            for (Element titleTag : titleTags) {
                d.title = titleTag.text();
            }

            Elements metaTags = doc.getElementsByTag("meta");
            for (Element metaTag : metaTags) {
                String content = metaTag.attr("content");
                if (TextUtils.isEmpty(content)) {
                    continue;
                }
                String name = metaTag.attr("name");
                if (TextUtils.isEmpty(name)) {
                    name = metaTag.attr("property");
                }

                switch (name) {
                    case "og:image":
                        d.imageUrl = content;
                        break;
                    case "og:title":
                        d.title = content;
                        break;
                }
            }

            return Response.success(d, HttpHeaderParser.parseCacheHeaders(response));
        }
    }

    @BindingAdapter({"ogImageUrl"})
    public static void loadOgImageUrl(ImageView v, OpenGraphLiveData.Data d) {
        if (TextUtils.isEmpty(d.imageUrl)) {
            v.setImageDrawable(ContextCompat.getDrawable(v.getContext(), R.drawable.og_image_default));
            return;
        }

        Picasso.get().load(d.imageUrl)
                .placeholder(R.drawable.og_image_default)
                .into(v);
    }

    @BindingAdapter({"ogTitle"})
    public static void loadOgTitle(TextView v, OpenGraphLiveData.Data d) {
        v.setText(d.title);
    }
}

File: app/src/main/java/com/daohoangson/ydls/viewmodel/OpenGraphLiveData.kt
package com.daohoangson.ydls.viewmodel

import android.app.Application
import android.os.Handler
import android.os.Looper
import android.text.TextUtils
import android.widget.TextView

import com.android.volley.toolbox.HttpHeaderParser
import com.daohoangson.ydls.R

import org.jsoup.Jsoup

import java.net.HttpURLConnection
import java.util.HashMap

import androidx.core.content.ContextCompat
import androidx.databinding.BindingAdapter
import androidx.lifecycle.LiveData
import androidx.lifecycle.Observer
import com.android.volley.*
import com.android.volley.toolbox.NetworkImageView
import com.daohoangson.ydls.Volley

class OpenGraphLiveData internal constructor(application: Application, private val mSourceMediaUrl: LiveData<String>) : LiveData<OpenGraphLiveData.Data>(), Observer<String> {

    private val mHandler: Handler = Handler(Looper.getMainLooper())
    private val mQueue: RequestQueue = Volley.getInstance(application.applicationContext).requestQueue

    init {
        // https://github.com/google/volley/issues/51
        HttpURLConnection.setFollowRedirects(true)

        value = Data(null)
    }

    override fun onActive() {
        super.onActive()

        mSourceMediaUrl.observeForever(this)
    }

    override fun onChanged(url: String) {
        mHandler.postDelayed({ parseUrl(url) }, 1000)
    }

    override fun onInactive() {
        super.onInactive()

        cancelQueue()
        mSourceMediaUrl.removeObserver(this)
    }

    private fun cancelQueue() {
        mQueue.cancelAll(TAG)
    }

    private fun parseUrl(url: String) {
        if (Looper.myLooper() != Looper.getMainLooper()) {
            throw IllegalThreadStateException()
        }

        if (url != mSourceMediaUrl.value) {
            // url is outdated (because of the delayed processing)
            return
        }

        val currentData = value
        if (currentData != null) {
            if (url == currentData.mediaUrl) {
                // url is already being processed
                return
            }

            if (currentData.isLoading) {
                cancelQueue()
            }
        }

        val loadingData = Data(url)
        loadingData.isLoading = true
        postValue(loadingData)

        val req = OgRequest(url)
        mQueue.add(req.setTag(TAG))
    }

    class Data internal constructor(internal val mediaUrl: String?) {
        internal var hasData = false
        internal var isLoading = false
        internal var title = ""
        internal var imageUrl = ""
    }

    internal inner class OgRequest(url: String) : Request<Data>(Request.Method.GET, url, Response.ErrorListener { postValue(Data(url)) }) {

        override fun deliverResponse(d: Data) {
            postValue(d)
        }

        override fun getHeaders(): Map<String, String> {
            val headers = HashMap<String, String>()

            headers["User-Agent"] = "libcurl"

            return headers
        }

        override fun parseNetworkResponse(response: NetworkResponse): Response<Data> {
            val html = String(response.data)
            val d = Data(url)
            d.hasData = true

            val doc = Jsoup.parse(html)
            val titleTags = doc.getElementsByTag("title")
            for (titleTag in titleTags) {
                d.title = titleTag.text()
            }

            val metaTags = doc.getElementsByTag("meta")
            for (metaTag in metaTags) {
                val content = metaTag.attr("content")
                if (TextUtils.isEmpty(content)) {
                    continue
                }
                var name = metaTag.attr("name")
                if (TextUtils.isEmpty(name)) {
                    name = metaTag.attr("property")
                }

                when (name) {
                    "og:image" -> d.imageUrl = content
                    "og:title" -> d.title = content
                }
            }

            return Response.success(d, HttpHeaderParser.parseCacheHeaders(response)
                    ?: buildDefaultCacheEntry(response))
        }

        private fun buildDefaultCacheEntry(response: NetworkResponse): Cache.Entry {
            val entry = Cache.Entry()

            entry.data = response.data
            entry.etag = url
            entry.responseHeaders = response.headers
            entry.allResponseHeaders = response.allHeaders

            val now = System.currentTimeMillis()
            entry.softTtl = now + 600000
            entry.ttl = entry.softTtl
            entry.serverDate = now
            entry.lastModified = now

            return entry
        }
    }

    companion object {
        private const val TAG = "OpenGraph"

        @BindingAdapter("ogImageUrl")
        @JvmStatic
        fun setOpenGraphDataImageUrl(v: NetworkImageView, d: OpenGraphLiveData.Data) {
            if (!d.hasData || TextUtils.isEmpty(d.imageUrl)) {
                v.setImageDrawable(ContextCompat.getDrawable(v.context, R.drawable.og_image_default))
                return
            }

            v.setImageUrl(d.imageUrl, Volley.getInstance(v.context).imageLoader)
        }

        @BindingAdapter("ogTitle")
        @JvmStatic
        fun setOpenGraphDataTitle(v: TextView, d: OpenGraphLiveData.Data) {
            v.text = d.title
        }
    }
}
